shared/idmap/shift_linux.go

```diff
@@ -156,7 +156,7 @@ int shiftowner(char *basepath, char *path, int uid, int gid)
 	char fdpath[PATH_MAX], realpath[PATH_MAX];
 	struct stat sb;
 
-	fd = open(path, O_PATH | O_NOFOLLOW);
+	fd = open(path, O_PATH | O_NOFOLLOW | O_CLOEXEC);
```

Comment:
Fwiw, would you mind changing all of them to `O_NOCTTY | O_NOFOLLOW | O_CLOEXEC)`?

Reply:
Like, the last path component of a loop device should definitely not be a symlink.

Reply:
I don't think we want to pass NOFOLLOW to all of them, there are many cases where we actually want to follow symlinks.
Adding `O_NOCTTY` is probably fine to do everywhere even though when using a fixed path, it's not really all that needed (mostly matters on user provided paths or when traversing a large tree like here).

I'd prefer we land the PR as it stands now so we can make sure it doesn't regress anything with shifting and fixes the issue we've hit in our jenkins tests, but please go ahead and open another PR to sprinkle O_NOCTTY where you think it makes sense!

Reply:
What no, you misunderstood. Not to all of our open calls to the two open calls in this pr.

Reply:
Hmm, why would we care if the loop device path is a symlink? It's not user provided so if the /dev on the system is somehow using weird symlinks, we should probably still follow those.

`NOFOLLOW | NOCTTY` should be set whenever we walk a tree or with some user provided paths.
So I think we need to go through every open call and make the decision there of whether one or both flag is needed to ensure we don't accidentally open the wrong thing.

Reply:
We _are_ following symlinks. In fact, there is no way to block symlink resolution in the path before `openat2()`. This is about the last component. And especially for well-known paths being as strict as possible makes even more sense.

For example, the changes here to `find_associated_loop_device()`. It looks up the backing file via `/sys/block/loop*/loop/backing_file`. The last component cannot be a symlink or a terminal so we should express this in our open call explicitly. So both `O_NOFOLLOW` and `O_NOCTTY` should be there.

For `shiftowner()` we already use `O_NOFOLLOW` since we're going through `/proc/self/fd/%d` and read the link contents out. But it also should absolutely be `O_NOCTTY` because if this is somehow a device file we're about to chown we don't want it to become our controlling terminal.

Reply:
Sounds good